Add notes for n25218 (#529)
bitcoin-core-review-club/website: _posts/2022-07-20-#25218.md

components: ["refactoring"]
host: stickies-v
status: upcoming
commit: 111ea3a
---

## Notes

> - To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful. With that pattern, one or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) can be used to access the data that was modified or created within the function, as well as to access information about how the operation happened (e.g. error messages).

Reviewer: ```suggestion
- To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful, along with one or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) to access the function result and information about the operation, e.g. error messages.
```

Author: Simplification here makes sense - just modified your suggestion a bit to not have "along with" to highlight that out-parameters are not being "returned".

> To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful. One or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) can then be used to access the function result and information about the operation (e.g. error messages).

- To put it in (pseudo)code, you'll find many functions that look somewhat like this:
```cpp
bool doSomething(arg1, arg2, arg3, arg4, &result_obj, &error_string) {
    // do something...
    if (error) {
        error_string = "something bad happened";
        return false;
    }

    result = goodResult;
    return true;
}

> ```

Author: Yes good idea, added the one provided in the PR

- Taking advantage of this new `BResult` class for existing code requires both the function as well all of its call sites to be refactored.
  
- The result of the kind of code simplification enabled by #25218 is quite visible in e.g. [`CWallet::GetNewDestination`](https://github.com/bitcoin/bitcoin/pull/25218/files#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L2310-L2327), where both the actual result as well as the error string are no longer out-parameters.

## Questions
1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

2. What are the different types of parameters and return values that the PR assumes we commonly use in a function signature? Which of these parameters are affected by the PR?

3. Why does [`BResult`](https://github.com/bitcoin/bitcoin/blob/111ea3ab711414236f8678566a7884d48619b2d8/src/util/result.h#L16) have a separate constructor `BResult(const bilingual_str& error)` that seems to do the exact same as the templated constructor `BResult(const T& _obj)`? Does this introduce any edge cases or limitations, and if so - are they documented?

> 4. Do you know of any other commonly used return type(s) that are similar to `BResult`?

Author: Fixed!

5. In commit `wallet: refactor, include 'FeeCalculation' inside 'CreatedTransactionResult'`, what do you think is the rationale for making `FeeCalculation` a member of `CreatedTransactionResult` instead of having it as an out-parameter in `CreateTransaction()`?

6. Should someone now follow up with a big PR that refactors all functions that would benefit from using `BResult`? Why (not)?

> 7. (Bonus) Do you know of any others functions that you think would benefit nicely from being refactored to use `BResult`?

Author: Fixed!
